[SPARK-27134][SQL] array_distinct function does not work correctly with columns containing array of array

dilipbiswal opened the pull request:
## What changes were proposed in this pull request?
Correct the logic to compute the distinct.

Below is a small repro snippet.

```
scala> val df = Seq(Seq(Seq(1, 2), Seq(1, 2), Seq(1, 2), Seq(3, 4), Seq(4, 5))).toDF("array_col")
df: org.apache.spark.sql.DataFrame = [array_col: array<array<int>>]

scala> val distinctDF = df.select(array_distinct(col("array_col")))
distinctDF: org.apache.spark.sql.DataFrame = [array_distinct(array_col): array<array<int>>]

scala> df.show(false)
+----------------------------------------+
|array_col                               |
+----------------------------------------+
|[[1, 2], [1, 2], [1, 2], [3, 4], [4, 5]]|
+----------------------------------------+
```
Error 
```
scala> distinctDF.show(false)
+-------------------------+
|array_distinct(array_col)|
+-------------------------+
|[[1, 2], [1, 2], [1, 2]] |
+-------------------------+
```
Expected result
```
scala> distinctDF.show(false)
+-------------------------+
|array_distinct(array_col)|
+-------------------------+
|[[1, 2], [3, 4], [4, 5]] |
+-------------------------+
```
## How was this patch tested?
Added an additional test.

dilipbiswal committed:
[SPARK-27134] array_distinct function does not work correctly with columns containing array of array

dilipbiswal committed:
remove space

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8811/
Test PASSed.

SparkQA commented:
**[Test build #103388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103388/testReport)** for PR 24073 at commit [`860ed87`](https://github.com/apache/spark/commit/860ed8700f6201e61fe8b6e293d34c7a27d4cfa4).

SparkQA commented:
**[Test build #103388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103388/testReport)** for PR 24073 at commit [`860ed87`](https://github.com/apache/spark/commit/860ed8700f6201e61fe8b6e293d34c7a27d4cfa4).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103388/
Test FAILed.

dilipbiswal committed:
style and one more test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8814/
Test PASSed.

SparkQA commented:
**[Test build #103391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103391/testReport)** for PR 24073 at commit [`049bf9b`](https://github.com/apache/spark/commit/049bf9b28072b8e6c862123a664d89762017c49c).

dilipbiswal committed:
variable names

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8818/
Test PASSed.

SparkQA commented:
**[Test build #103395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103395/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).

SparkQA commented:
**[Test build #103391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103391/testReport)** for PR 24073 at commit [`049bf9b`](https://github.com/apache/spark/commit/049bf9b28072b8e6c862123a664d89762017c49c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103391/
Test PASSed.

dilipbiswal commented:
cc @ushin @kiszk

SparkQA commented:
**[Test build #103395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103395/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103395/
Test FAILed.

dilipbiswal commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8831/
Test PASSed.

SparkQA commented:
**[Test build #103409 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103409/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).

SparkQA commented:
**[Test build #103409 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103409/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

dilipbiswal commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103409/
Test FAILed.

SparkQA commented:
**[Test build #103419 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103419/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8840/
Test PASSed.

maropu commented:
Can you put an example query to reproduce this bug in the PR description?

kiszk commented:
cc @ueshin

kiszk commented:
Thanks, let me see later.

dilipbiswal commented:
@maropu Updated the PR description. Thank you.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
           }
         }
       }
-      new GenericArrayData(data.slice(0, pos))
```
Looks like original implementation assumes the duplicate items are placed at the end of `data`? I think it not only affects array of array, but also other element type like `BinaryType`.

viirya replied:
This test case fails at current master, but passes after this change.

```scala
val a8 = Literal.create(Seq(2, 1, 2, 3, 4, 4, 5).map(_.toString.getBytes), ArrayType(BinaryType))
checkEvaluation(new ArrayDistinct(a8), Seq(2, 1, 3, 4, 5).map(_.toString.getBytes))
```

dilipbiswal replied:
@viirya 
>Looks like original implementation assumes the duplicate items are placed at the end of data?
Probably. The thing is, it does not rearrange the data in any way. So i don't know how we can just return the slice of the original array.
> I think it not only affects array of array, but also other element type like BinaryType
Yeah.. I will add your test case or enhance the Array[Binary] test case.

ueshin replied:
Good catch.
Actually I'm not sure how I could miss this.
Thanks!

SparkQA commented:
**[Test build #103419 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103419/testReport)** for PR 24073 at commit [`d59b0f7`](https://github.com/apache/spark/commit/d59b0f70d6844b9b763fcd7de77309fdd18658d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103419/
Test PASSed.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
+          var j = 0;
+          while (!found && j < arrayBuffer.size) {
+            val va = arrayBuffer(j)
+            found = (va != null) && ordering.equiv(va, data(i))
```
Rather than handle nulls separately, can you just check it here?
```
found = if (va == null) data(i) == null else ordering.equiv(...)
```
It also kind of looks like the ordering already handles nulls?

dilipbiswal replied:
@srowen Thanks. Actually in my understanding, ordering does not seem to handle nulls.  Given that, the proposed condition does not work  for case when va != null and data[i] is null. We get a null pointer exception. We can probably do something like : 
```
 if (data(i) != null && va != null) {
     found = ordering.equiv(va, data(i))
  } else if (data(i) == null && va == null) {
     found = true
  }
```
Given this, i feel the existing code reads better and i think performs better if we have many null and non-null values in the array by keeping the null handling outside the inner loop. Also, i believe, for other collection operation functions we treat nulls separately. But i will change, if you feel otherwise. Please let me know.

srowen replied:
You're right, on second look the ordering in ArrayType doesn't handle nulls, only nulls in the array. This is fine. You could save a lookup with...

```
if (data(i) == null) {
  found = va == null
} else if (va != null) {
  found = ordering.equiv(va, data(i))
}
```

but I don't feel strongly about it

kiszk replied:
@dilipbiswal explained my intention in the code of `null` part. @srowen is code simple and easy for reading, but it may include # of iterations if we have already seen `null`.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
-            pos = pos + 1
+        if (data(i) != null) {
+          found = false
+          var j = 0;
```
Remove semicolon

dilipbiswal replied:
@srowen will do.

srowen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
-      var pos = 0
+      val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+      var alreadyStoredNull = false
+      var found = false
```
Move this var inside the if statement below?

dilipbiswal replied:
@swowen ok.

dilipbiswal commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala:
```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
     (data: Array[AnyRef]) => {
-      var foundNullElement = false
-      var pos = 0
+      val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
```
We are returning a GenericArrayData(arrayBuffer). GenericArrayData takes a parameter of Array[Any]. So we are okay here, no ?

srowen replied:
That's fine, you can pass an Array of anything to it then (right? or is there some compile-time issue I'm not thinking of). It's not that it doesn't work but this code locally could be more precise. No big deal.

dilipbiswal replied:
@srowen Sorry.. a little confused. So we have a input which is a Array[AnyRef]. Now if we declare the temporary buffer a ArrayBuffer[Array[AnyRef]], how do we populate its content ?
Example :
Input1 : Array[Integer] => Seq(1, 2, , 1)
In this case our output is : ArrayBuffer[Int] = Array(1, 2)
Input2 : Array[Array[Integer]] => Seq(Seq(1, 2), Seq(3, 4), Seq(3,4))
In this case our output is : ArrayBuffer[Array[Int]] => Array(Array(1,2), Array(3,4))
Input3 : Array[Struct] => Seq(struct(...), struct(...))

Thanks for your help.

srowen replied:
Disregard this, I'm mistaken. The use case here was arrays of arrays but this code isn't handling only array elements. Can the type by AnyRef though?

dilipbiswal replied:
@srowen Thank you.. Sure i will change it to Anyref.

dilipbiswal committed:
Code review

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8885/
Test PASSed.

SparkQA commented:
**[Test build #103467 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103467/testReport)** for PR 24073 at commit [`dfdd109`](https://github.com/apache/spark/commit/dfdd109e909d4176a715870f2dbc9e617de11973).

dilipbiswal commented:
@HyukjinKwon Hello.. the test seems to be stuck in `ApproxCountDistinctForIntervalsQuerySuite` which seems unrelated to the change in this PR. Is there a way to stop/kill this run ?

maropu commented:
Jenkins has weird behaivours now and it seems the tests stopped in https://github.com/apache/spark/pull/24028#issuecomment-472554054, too.

maropu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8895/
Test PASSed.

SparkQA commented:
**[Test build #103479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103479/testReport)** for PR 24073 at commit [`dfdd109`](https://github.com/apache/spark/commit/dfdd109e909d4176a715870f2dbc9e617de11973).

SparkQA commented:
**[Test build #103479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103479/testReport)** for PR 24073 at commit [`dfdd109`](https://github.com/apache/spark/commit/dfdd109e909d4176a715870f2dbc9e617de11973).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103479/
Test FAILed.

dilipbiswal commented:
retest this please

SparkQA commented:
**[Test build #103482 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103482/testReport)** for PR 24073 at commit [`dfdd109`](https://github.com/apache/spark/commit/dfdd109e909d4176a715870f2dbc9e617de11973).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8899/
Test PASSed.

SparkQA commented:
**[Test build #103482 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103482/testReport)** for PR 24073 at commit [`dfdd109`](https://github.com/apache/spark/commit/dfdd109e909d4176a715870f2dbc9e617de11973).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103482/
Test PASSed.

dilipbiswal committed:
Code review

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8928/
Test PASSed.

SparkQA commented:
**[Test build #103515 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103515/testReport)** for PR 24073 at commit [`60eb195`](https://github.com/apache/spark/commit/60eb195656119181a64644148d4499eacaa25e0c).

SparkQA commented:
**[Test build #103515 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103515/testReport)** for PR 24073 at commit [`60eb195`](https://github.com/apache/spark/commit/60eb195656119181a64644148d4499eacaa25e0c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103515/
Test PASSed.

ueshin commented:
LGTM.
@srowen Could you take another look please?
Thanks!

dilipbiswal committed:
Revert "Code review"

This reverts commit 60eb195656119181a64644148d4499eacaa25e0c.

dilipbiswal commented:
@kiszk @ueshin @srowen Thanks a lot for reviewing. I have reverted the last commit where i had removed special casing null handling. It should be ok now. Thanks.

srowen reviewed (approved):
This seems fine. Is there a more general problem of this form? someone else mentioned it might not be specific to arrays. It's OK to consider that separately if so, just checking if there is another very similar fix to be made elsewhere

kiszk commented:
LGTM, thank you for fixing my mistakes.

ueshin commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8948/
Test PASSed.

SparkQA commented:
**[Test build #103537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103537/testReport)** for PR 24073 at commit [`1bde41c`](https://github.com/apache/spark/commit/1bde41c49528b0b58329f07fd3daef84fa411081).

SparkQA commented:
**[Test build #103537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103537/testReport)** for PR 24073 at commit [`1bde41c`](https://github.com/apache/spark/commit/1bde41c49528b0b58329f07fd3daef84fa411081).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103537/
Test PASSed.

srowen commented:
Merged to master/2.4. It didn't pick cleanly into 2.3 and I wasn't clear whether it affected 2.3

dilipbiswal commented:
Thanks a lot @viirya @ueshin @ueshin @kiszk

dilipbiswal commented:
@srowen Actually this function was added in 2.4. So we should be good :-)
